src/RustyObjectStore.jl

```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
+"""
+struct ClientOptions
+    params::Dict{String, String}
+
+    function ClientOptions(;
+        request_timeout_secs::Option{Int} = nothing,
+        connect_timeout_secs::Option{Int} = nothing,
+        max_retries::Option{Int} = nothing,
+        retry_timeout_secs::Option{Int} = nothing
+    )
+        params = Dict()
+        if !isnothing(request_timeout_secs)
+            params["timeout"] = "$(string(request_timeout_secs))s"
+        end
+
+        if !isnothing(connect_timeout_secs)
+            params["connect_timeout"] = "$(string(connect_timeout_secs))s"
+        end
+
+        if !isnothing(max_retries)
+            params["max_retries"] = string(max_retries)
+        end
+
+        if !isnothing(retry_timeout_secs)
+            params["retry_timeout_secs"] = string(retry_timeout_secs)
+        end
+
+        return new(params)
+    end
+end
+
+abstract type AsConfig end
+
+function as_config(wrapper::AsConfig) end
+
 """
     $TYPEDEF
 
 # Arguments
-$TYPEDFIELDS
+- `url::String`: Url of the object store container root path.
+  It must include the cloud specific url scheme (s3://, azure:// ...).
+- `params::Dict{String, String}`: A set of key-value pairs to configure access to the object store.
+  Refer to the object_store crate documentation for the list of all supported parameters.
 """
-struct AzureCredentials
-    "Azure account name"
-    account::String
-    "Azure container name"
-    container::String
-    "Azure access key"
-    key::String
-    "(Optional) Alternative Azure host. For example, if using Azurite."
-    host::String
-    function AzureCredentials(account::String, container::String, key::String, host::String="")
-        return new(account, container, key, host)
+struct Config <: AsConfig
+    config_string::String
+    function Config(url::String, params::Dict{String, String})
+        dict = merge(Dict("url" => url), params)
+        config_string = JSON3.write(dict)
+        new(config_string)
     end
 end
-function Base.show(io::IO, credentials::AzureCredentials)
-    print(io, "AzureCredentials("),
-    print(io, repr(credentials.account), )
-    print(io, ", ", repr(credentials.container))
-    print(io, ", ", "\"*****\"") # don't print the secret key
-    !isempty(credentials.host) && print(io, ", ", repr(credentials.host))
+
+as_config(wrapper::Config) = wrapper
+
+"""
+    $TYPEDEF
+
+# Keyword Arguments
+- `storage_account_name::String`: Azure storage account name.
+- `container_name::String`: Azure container name.
+- `storage_account_key::Option{String}`: (Optional) Azure storage account key.
+- `storage_sas_token::Option{String}`: (Optional) Azure storage SAS token.
+- `host::Option{String}`: (Optional) Alternative Azure host. For example, if using Azurite.
+- `opts::ClientOptions`: (Optional) Client configuration options.
+"""
+struct AzureConfig <: AsConfig
+    config::Config
```

Comment:
Why store this as `Config` rather than just a `String`? Does it make FFI cleaner somehow to guarantee they both go through this type? 

I wonder if `Config` could instead just be a function, `config_string(params...)`, that calls `JSON3.write`?

Reply:
I was assuming that we would want to keep the `cconvert` and `unsafe_convert` implementations limited to a single concrete struct. But thinking about it, we could make those functions operate on the supertype directly and then make do with a string field directly on each of the config types (`Config`, `AzureConfig`, `AWSConfig`). For efficiency we don't want to keep serializing the JSON on each native call, that is why we need a string field on a struct.

Would that be better than wrapping the `Config` type